src/hostcalls.rs

```diff
@@ -1199,48 +1356,189 @@ mod utils {
+                return map; // 转换失败，返回空的 Vec
+            },
+        };
+
+        // 计算出初始偏移量 `p`
         let mut p = 4 + size * 8;
+
+        // 遍历每个键值对
         for n in 0..size {
             let s = 4 + n * 8;
-            let size = u32::from_le_bytes(<[u8; 4]>::try_from(&bytes[s..s + 4]).unwrap()) as usize;
-            let key = bytes[p..p + size].to_vec();
-            p += size + 1;
-            let size =
-                u32::from_le_bytes(<[u8; 4]>::try_from(&bytes[s + 4..s + 8]).unwrap()) as usize;
-            let value = bytes[p..p + size].to_vec();
-            p += size + 1;
-            map.push((
-                String::from_utf8(key).unwrap(),
-                String::from_utf8(value).unwrap(),
-            ));
+
+            // 确保有足够的数据来处理当前条目
+            if bytes.len() < s + 8 || bytes.len() < p {
+                error!("Insufficient bytes to read key-value pair.");
+                return map;
+            }
+
+            // 读取 key 的大小
+            let key_size = match <[u8; 4]>::try_from(&bytes[s..s + 4]) {
+                Ok(byte_array) => u32::from_le_bytes(byte_array) as usize,
+                Err(_) => {
+                    error!("Failed to convert bytes to array when extracting key size.");
+                    return map;
+                },
+            };
+
+            // 确保有足够的字节用于 key
+            if bytes.len() < p + key_size {
+                error!("Insufficient bytes to read the key.");
+                return map;
+            }
+
+            // 提取 key 的字节数组
+            let key = bytes[p..p + key_size].to_vec();
+            p += key_size + 1; // 跳过 key 后的 '\0' 字节
+
+            // 读取 value 的大小
+            let value_size = match <[u8; 4]>::try_from(&bytes[s + 4..s + 8]) {
+                Ok(byte_array) => u32::from_le_bytes(byte_array) as usize,
+                Err(_) => {
+                    error!("Failed to convert bytes to array when extracting value size.");
+                    return map;
+                },
+            };
+
+            // 确保有足够的字节用于 value
+            if bytes.len() < p + value_size {
+                error!("Insufficient bytes to read the value.");
+                return map;
+            }
+
+            // 提取 value 的字节数组
+            let value = bytes[p..p + value_size].to_vec();
+            p += value_size + 1; // 跳过 value 后的 '\0' 字节
+
+            // 尝试将 key 和 value 转换为 String，处理转换失败的情况
+            let key_string = String::from_utf8(key).unwrap_or_else(|e| {
+                error!("Failed to convert key to String: {}", e);
+                String::new() // 或者根据你的需求返回默认值
+            });
+
+            let value_string = String::from_utf8(value).unwrap_or_else(|e| {
+                error!("Failed to convert value to String: {}", e);
+                String::new() // 或者根据你的需求返回默认值
+            });
+
+            // 将成功转换的键值对推入 map
+            map.push((key_string, value_string));
```

Comment:
Map is serialized by a trusted host environment, so it's not untrusted/malicious input. See: #282.